# Trabalho de ICD: UAP (Fenômenos Anômalos não Identificados)
*Professor: Yuri Malheiros Alunos: Vitoria, Simplicio, Daiellem, Bruno*

### Pré-Processamento: Limpeza e transformação de Dados

In [120]:
import pandas as pd
import datetime as dt
import numpy as np

In [121]:
df1 = pd.read_csv("./ufo_data_nuforc.csv")
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1317 entries, 0 to 1316
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   posted      1317 non-null   object 
 1   date        1317 non-null   object 
 2   time        1317 non-null   object 
 3   city        1317 non-null   object 
 4   state       1317 non-null   object 
 5   shape       1317 non-null   object 
 6   duration    1243 non-null   object 
 7   summary     1317 non-null   object 
 8   images      1317 non-null   object 
 9   img_link    1317 non-null   object 
 10  lat         1317 non-null   float64
 11  lng         1317 non-null   float64
 12  population  1317 non-null   int64  
dtypes: float64(2), int64(1), object(10)
memory usage: 133.9+ KB


O primeiro [DataSet1](https://www.kaggle.com/datasets/joebeachcapital/ufo-sightings)tem  9 anos de registros, de 2014 a 2023, possui 1317 linhas e 13 colunas. As colunas têm informações cronológicas(data e hora), geográficas(cidade, estado, latitude, longitude e população), morfologia dos UAP(forma), detalhes das observações(duração, sumario, imagens e link das imagens).
Vamos fazer a verificaçãp de linhas duplicadas:

In [17]:
df1 = df1.drop_duplicates()

A coluna date conta com as datas das observações, como futuramente podemos precisar observar os dados por ano, mes ou dia vamos separar essas datas e guarda em outras coluna. Primeiro transformando a coluna para o tipo datetime:

In [22]:
df1['date'] =  pd.to_datetime(df1['date'])

Agora conseguimos separar esses dados em novas colunas:

In [19]:
df1['year']=df1['date'].dt.year

In [23]:
df1['month']=df1['date'].dt.month

In [24]:
df1['day']=df1['date'].dt.day

Agora nós temos as datas separadas e podemos remover essa coluna do dataset

In [25]:
df1 = df1.drop(["date"], axis = 1)

Vamos fazer o mesmo processo com as horas.

In [114]:
df1['time'] =  pd.to_datetime(df1['time'])

In [116]:
df1['hour']= df1['time'].dt.hour

In [117]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1317 entries, 0 to 1316
Data columns (total 14 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   posted      1317 non-null   object        
 1   date        1317 non-null   object        
 2   time        1317 non-null   datetime64[ns]
 3   city        1317 non-null   object        
 4   state       1317 non-null   object        
 5   shape       1317 non-null   object        
 6   duration    1243 non-null   object        
 7   summary     1317 non-null   object        
 8   images      1317 non-null   object        
 9   img_link    1317 non-null   object        
 10  lat         1317 non-null   float64       
 11  lng         1317 non-null   float64       
 12  population  1317 non-null   int64         
 13  hour        1317 non-null   int32         
dtypes: datetime64[ns](1), float64(2), int32(1), int64(1), object(9)
memory usage: 139.0+ KB


Pronto. Agora é só salvar o dataset com as alterações feitas.

In [118]:
df1.to_csv('df1.csv')

In [29]:
df2= pd.read_csv("./datasetETs.csv", low_memory=False)

Date_time                  city state/province country  \
0      10/10/1949 20:30            san marcos             tx      us   
1      10/10/1949 21:00          lackland afb             tx     NaN   
2      10/10/1955 17:00  chester (uk/england)            NaN      gb   
3      10/10/1956 21:00                  edna             tx      us   
4      10/10/1960 20:00               kaneohe             hi      us   
...                 ...                   ...            ...     ...   
80327    9/9/2013 21:15             nashville             tn      us   
80328    9/9/2013 22:00                 boise             id      us   
80329    9/9/2013 22:00                  napa             ca      us   
80330    9/9/2013 22:20                vienna             va      us   
80331    9/9/2013 23:00                edmond             ok      us   

      UFO_shape length_of_encounter_seconds described_duration_of_encounter  \
0      cylinder                        2700                      45 minutes   
1         light                        7200                         1-2 hrs   
2        circle                          20                      20 seconds   
3        circle                          20                        1/2 hour   
4         light                         900                      15 minutes   
...         ...                         ...                             ...   
80327     light                         600                      10 minutes   
80328    circle                        1200                      20 minutes   
80329     other                        1200                            hour   
80330    circle                           5                       5 seconds   
80331     cigar                        1020                      17 minutes   

                                             description date_documented  \
0      This event took place in early fall around 194...       4/27/2004   
1      1949 Lackland AFB&#44 TX.  Lights racing acros...      12/16/2005   
2      Green/Orange circular disc over Chester&#44 En...       1/21/2008   
3      My older brother and twin sister were leaving ...       1/17/2004   
4      AS a Marine 1st Lt. flying an FJ4B fighter/att...       1/22/2004   
...                                                  ...             ...   
80327  Round from the distance/slowly changing colors...       9/30/2013   
80328  Boise&#44 ID&#44 spherical&#44 20 min&#44 10 r...       9/30/2013   
80329                                       Napa UFO&#44       9/30/2013   
80330  Saw a five gold lit cicular craft moving fastl...       9/30/2013   
80331  2 witnesses 2  miles apart&#44 Red &amp; White...       9/30/2013   

         latitude   longitude  
0      29.8830556  -97.941111  
1        29.38421  -98.581082  
2            53.2   -2.916667  
3      28.9783333  -96.645833  
4      21.4180556 -157.803611  
...           ...         ...  
80327  36.1658333  -86.784444  
80328  43.6136111 -116.202500  
80329  38.2972222 -122.284444  
80330  38.9011111  -77.265556  
80331  35.6527778  -97.477778  

[80332 rows x 11 columns]

Nosso segundo [dataset](https://www.kaggle.com/datasets/camnugent/ufo-sightings-around-the-world) com uma rápida olhada percebemos que a coluna da latitude que conta com dados numéricos está como object, muitas das vezes isso acontece quando digitamos a virgula em vez do ponto, mas não é esse o problema. Se transformarmos a coluna direto para float pode apresentar outro problema bem maior, então vamos aplicar a coluna em algum comando para ele nos mostrar qual o problema. Descobrimos que foi digitado, em uma das linhas, uma letra entre dos números. Sabendo qual a letra podemos usar replace para removela e transformar o tipo da coluna para float.

In [130]:
df2["latitude"] = df2["latitude"].str.replace('q', '').astype(float)

In [131]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80332 entries, 0 to 80331
Data columns (total 14 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   city                             80332 non-null  object 
 1   state/province                   74535 non-null  object 
 2   country                          70662 non-null  object 
 3   UFO_shape                        78400 non-null  object 
 4   length_of_encounter_seconds      80332 non-null  object 
 5   described_duration_of_encounter  80332 non-null  object 
 6   description                      80317 non-null  object 
 7   date_documented                  80332 non-null  object 
 8   latitude                         80332 non-null  float64
 9   longitude                        80332 non-null  float64
 10  year                             80332 non-null  int32  
 11  month                            80332 non-null  int32  
 12  day               

Agora vamos fazer o mesmo processo do datetime como no outro datasete, só que esse está com datas e horas:

In [83]:
df2['Date_time']= pd.to_datetime(df2['Date_time'], format= 'mixed', utc=True)

DateParseError: hour must be in 0..23: 10/11/2006 24:00, at position 388

Nos deparamos com esse erro. Nós podemos corrigir isso aumentado uma hora ou diminuido, optamos por diminuir

In [96]:
df2.loc[388, 'Date_time'] = '10/11/2006 23:00'

compilamos outras vez e apareceu mais um erro desse, então usamos o replace para corrigir toda a coluna:

In [100]:
df2["Date_time"] = df2["Date_time"].str.replace('24:00', '23:00')

Agora conseguimos mudar o tipo:

In [101]:
df2['Date_time']= pd.to_datetime(df2['Date_time'], format= 'mixed')

Vamos separar os dados em novas colunas e em seguida remover a coluna Date_time:

In [104]:
df2['year']=df2['Date_time'].dt.year

In [105]:
df2['month']=df2['Date_time'].dt.month

In [106]:
df2['day']=df2['Date_time'].dt.day

In [107]:
df2['hour']=df2['Date_time'].dt.hour

In [108]:
df2

Date_time                  city state/province country  \
0     1949-10-10 20:30:00            san marcos             tx      us   
1     1949-10-10 21:00:00          lackland afb             tx     NaN   
2     1955-10-10 17:00:00  chester (uk/england)            NaN      gb   
3     1956-10-10 21:00:00                  edna             tx      us   
4     1960-10-10 20:00:00               kaneohe             hi      us   
...                   ...                   ...            ...     ...   
80327 2013-09-09 21:15:00             nashville             tn      us   
80328 2013-09-09 22:00:00                 boise             id      us   
80329 2013-09-09 22:00:00                  napa             ca      us   
80330 2013-09-09 22:20:00                vienna             va      us   
80331 2013-09-09 23:00:00                edmond             ok      us   

      UFO_shape length_of_encounter_seconds described_duration_of_encounter  \
0      cylinder                        2700                      45 minutes   
1         light                        7200                         1-2 hrs   
2        circle                          20                      20 seconds   
3        circle                          20                        1/2 hour   
4         light                         900                      15 minutes   
...         ...                         ...                             ...   
80327     light                         600                      10 minutes   
80328    circle                        1200                      20 minutes   
80329     other                        1200                            hour   
80330    circle                           5                       5 seconds   
80331     cigar                        1020                      17 minutes   

                                             description date_documented  \
0      This event took place in early fall around 194...       4/27/2004   
1      1949 Lackland AFB&#44 TX.  Lights racing acros...      12/16/2005   
2      Green/Orange circular disc over Chester&#44 En...       1/21/2008   
3      My older brother and twin sister were leaving ...       1/17/2004   
4      AS a Marine 1st Lt. flying an FJ4B fighter/att...       1/22/2004   
...                                                  ...             ...   
80327  Round from the distance/slowly changing colors...       9/30/2013   
80328  Boise&#44 ID&#44 spherical&#44 20 min&#44 10 r...       9/30/2013   
80329                                       Napa UFO&#44       9/30/2013   
80330  Saw a five gold lit cicular craft moving fastl...       9/30/2013   
80331  2 witnesses 2  miles apart&#44 Red &amp; White...       9/30/2013   

         latitude   longitude  year  month  day  hour  
0      29.8830556  -97.941111  1949     10   10    20  
1        29.38421  -98.581082  1949     10   10    21  
2            53.2   -2.916667  1955     10   10    17  
3      28.9783333  -96.645833  1956     10   10    21  
4      21.4180556 -157.803611  1960     10   10    20  
...           ...         ...   ...    ...  ...   ...  
80327  36.1658333  -86.784444  2013      9    9    21  
80328  43.6136111 -116.202500  2013      9    9    22  
80329  38.2972222 -122.284444  2013      9    9    22  
80330  38.9011111  -77.265556  2013      9    9    22  
80331  35.6527778  -97.477778  2013      9    9    23  

[80332 rows x 15 columns]

In [109]:
df2 = df2.drop(["Date_time"], axis = 1)

In [133]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 80332 entries, 0 to 80331
Data columns (total 14 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   city                             80332 non-null  object 
 1   state/province                   74535 non-null  object 
 2   country                          70662 non-null  object 
 3   UFO_shape                        78400 non-null  object 
 4   length_of_encounter_seconds      80332 non-null  object 
 5   described_duration_of_encounter  80332 non-null  object 
 6   description                      80317 non-null  object 
 7   date_documented                  80332 non-null  object 
 8   latitude                         80332 non-null  float64
 9   longitude                        80332 non-null  float64
 10  year                             80332 non-null  int32  
 11  month                            80332 non-null  int32  
 12  day               

In [132]:
df2.to_csv('df2.csv')

Salvamos o dataset com as modificações feitas.

In [124]:
df3= pd.read_csv("./consolidated_weather_V03.csv")
df3

state             city  colour      shape  mday  month  year  hour  \
0        WA          Olympia     NaN  Formation    14     11  1997    21   
1        WA          Olympia     NaN   Fireball    14     11  1997    21   
2        WA          Olympia     NaN  Formation    14     11  1997    21   
3        MD        Salisbury     NaN       Oval    24     11  1997     2   
4        MD        Salisbury  YELLOW       Oval    24     11  1997     2   
...     ...              ...     ...        ...   ...    ...   ...   ...   
51865    AZ  Prescott Valley     NaN     Circle    28      1  2000    20   
51866    CA           Covina     NaN      Light    27      1  2000    18   
51867    WA          Seattle     NaN   Fireball    26      1  2000    19   
51868    TX         Edinburg     NaN      Light    25      1  2000    19   
51869    NM       Rio Rancho     NaN     Sphere    23      1  2000    15   

            id  pressure  ...     dewpt  thunder  fog    precip      wspd  \
0          NaN -0.163149  ... -0.867472        0    0       NaN -0.357641   
1          NaN -0.163149  ... -0.867472        0    0       NaN -0.357641   
2          NaN -0.163149  ... -0.867472        0    0       NaN -0.357641   
3          NaN -0.189379  ... -0.464073        0    0       NaN -0.911239   
4          NaN -0.189379  ... -0.464073        0    0       NaN -0.911239   
...        ...       ...  ...       ...      ...  ...       ...       ...   
51865  57531.0  0.610616  ... -1.835629        0    0       NaN -0.357641   
51866  57536.0  0.413896  ...       NaN        0    0       NaN -0.911239   
51867  57543.0  1.751592  ... -0.595177        0    0       NaN -0.911239   
51868  57544.0  0.230291  ...  0.635190        0    0       NaN  0.373899   
51869  57549.0  0.545043  ... -2.481068        0    0 -0.092084  0.917612   

       tornado       hum  snow     wgust          conds  
0            0  1.598872     0       NaN          Clear  
1            0  1.598872     0       NaN          Clear  
2            0  1.598872     0       NaN          Clear  
3            0  0.980525     0       NaN  Mostly Cloudy  
4            0  0.980525     0       NaN  Mostly Cloudy  
...        ...       ...   ...       ...            ...  
51865        0 -0.731823     0       NaN          Clear  
51866        0       NaN     0       NaN          Clear  
51867        0  0.790264     0       NaN  Mostly Cloudy  
51868        0  0.171916     0       NaN       Overcast  
51869        0 -2.539302     0 -0.640525          Clear  

[51870 rows x 27 columns]

Identificamos que nosso [dataset](https://www.kaggle.com/datasets/emorelli/consolidated-ufo-weather-data) tem muitas colunas especifica para esses dados. Resolvemos então remove-las do nosso dataset. 

In [125]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51870 entries, 0 to 51869
Data columns (total 27 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   state      51870 non-null  object 
 1   city       51870 non-null  object 
 2   colour     80 non-null     object 
 3   shape      51868 non-null  object 
 4   mday       51870 non-null  int64  
 5   month      51870 non-null  int64  
 6   year       51870 non-null  int64  
 7   hour       51870 non-null  int64  
 8   id         51631 non-null  float64
 9   pressure   51659 non-null  float64
 10  temp       51544 non-null  float64
 11  hail       51870 non-null  int64  
 12  heatindex  3589 non-null   float64
 13  windchill  8718 non-null   float64
 14  rain       51870 non-null  int64  
 15  vis        51170 non-null  float64
 16  wdire      51863 non-null  object 
 17  dewpt      51393 non-null  float64
 18  thunder    51870 non-null  int64  
 19  fog        51870 non-null  int64  
 20  precip

In [126]:
df3 = df3.drop(["snow", "tornado", "fog", "thunder", "rain", "hail"], axis = 1)

In [127]:
df3 = df3.drop_duplicates()

Depois da remover as colunas, removemos as linhas duplicadas. 

In [128]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 51830 entries, 0 to 51869
Data columns (total 21 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   state      51830 non-null  object 
 1   city       51830 non-null  object 
 2   colour     70 non-null     object 
 3   shape      51828 non-null  object 
 4   mday       51830 non-null  int64  
 5   month      51830 non-null  int64  
 6   year       51830 non-null  int64  
 7   hour       51830 non-null  int64  
 8   id         51631 non-null  float64
 9   pressure   51619 non-null  float64
 10  temp       51504 non-null  float64
 11  heatindex  3583 non-null   float64
 12  windchill  8704 non-null   float64
 13  vis        51130 non-null  float64
 14  wdire      51823 non-null  object 
 15  dewpt      51354 non-null  float64
 16  precip     2010 non-null   float64
 17  wspd       51483 non-null  float64
 18  hum        51353 non-null  float64
 19  wgust      3272 non-null   float64
 20  conds      

In [129]:
df3.to_csv('df3.csv')

Salvamos nosso dataset com as alterações feitas.